### ISSUE TITLE
output resolved config with the checkpoint

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#1968]
currently we dont have the resolved config with the ckpt. We should, for reproducibility.

**Comment by Ankur-singh** [comment]
@felipemello1  I would like to give this a try. Can you please give more context? What do you mean by "resolved config"?

**Comment by RdoubleA** [comment]
The resolved config object is what is ultimately passed to the recipe and includes all the command line overrides and any interpolations resolved: https://github.com/pytorch/torchtune/blob/main/recipes/full_finetune_distributed.py#L905

To include it in the checkpoint, one approach could be to update the `TrainingProgress` class to include the config: https://github.com/pytorch/torchtune/blob/main/recipes/full_finetune_distributed.py#L865. Or you could include it in the checkpoint dict itself since it's a static object that won't change through the training run: https://github.com/pytorch/torchtune/blob/890deab3029eef65f94cedb37fda14479f65f129/torchtune/training/checkpointing/_checkpoint_client.py#L176

There is also the issue that if certain fields are not present in the cfg then we assign defautl values (via `cfg.get("field", <default>)`) and this won't be reflected in the final object. But this is a separate discussion that's worth having.

**Comment by felipemello1** [comment]
thanks @RdoubleA !

@Ankur-singh , one of our metric loggers saves the config, for example. In this issue i am proposing that this should happen regardless of the metric logger: https://github.com/pytorch/torchtune/blob/890deab3029eef65f94cedb37fda14479f65f129/torchtune/training/metric_logging.py#L232

**Comment by Ankur-singh** [comment]
Thanks @felipemello1 & @RdoubleA for the helping me understand the issue better. 

I would love to give it a try. Any pointers to help me get started?

**Comment by felipemello1** [comment]
Awesome, I think that the example in the wandb code should make it trivial. The issue is more how to do it elegantly and where to put this code, i.e. in the checkpointer or in the loggers.

1) I am thinking that this logic should be part of every logger, e.g. repeating the code or having some utility function `save_config`. I prefer the utility function. The config should be saved to output_dir, like in the wandb example.

2) If that makes sense to you, then the next step would be to create a simple PR showing this being added to `torchtune.training.metric_logging.DiskLogger`, which is our standard logger. 

3) If people like how it looks, then, you should update the PR to add it to every logger.

4) For the testing section, you can create a every simple config, like llama 1b training for 10 steps, and run each logger once and confirm that a new config is correctly saved.

@ebsmothers , does this sound ok?
